packages/gatsby-plugin-twitter/src/gatsby-browser.js

```diff
@@ -1,47 +1,69 @@
+  window.twttr = (function(d, s, id) {
+    var js,
+      fjs = d.getElementsByTagName(s)[0],
+      t = window.twttr || {}
+    if (d.getElementById(id)) return t
+    js = d.createElement(s)
+    js.id = id
+    js.src = `https://platform.twitter.com/widgets.js`
+    fjs.parentNode.insertBefore(js, fjs)
 
-    s.type = `text/javascript`
-    s.innerText = jsCode
-    document.getElementsByTagName(`head`)[0].appendChild(s)
-  }
-  addJS(`
-          window.twttr = (function(d, s, id) {
-  var js, fjs = d.getElementsByTagName(s)[0],
-    t = window.twttr || {};
-  if (d.getElementById(id)) return t;
-  js = d.createElement(s);
-  js.id = id;
-  js.src = "https://platform.twitter.com/widgets.js";
-  fjs.parentNode.insertBefore(js, fjs);
-
-  t._e = [];
-  t.ready = function(f) {
-    t._e.push(f);
-  };
-
-  return t;
-}(document, "script", "twitter-wjs"));
-`)
+    t._e = []
+    t.ready = function(f) {
+      t._e.push(f)
+    }
+
+    return t
+  })(document, `script`, `twitter-wjs`)
 }
 
-let injectedTwitterScript = false
-exports.onRouteUpdate = function({ location }) {
+let renderTweets = false
+const renderedTweets = []
+
+exports.onRouteUpdate = function() {
+  renderTweets = true
+  const tweets = document.querySelectorAll(`.twitter-tweet`)
+
   // If there's an embedded tweet, lazy-load the twitter script (if it hasn't
-  // already been loaded), and then run the twitter load function.
-  if (document.querySelector(`.twitter-tweet`) !== null) {
-    if (!injectedTwitterScript) {
-      injectTwitterScript()
-      injectedTwitterScript = true
-    }
+  // already been loaded), and then render the tweets.
+  if (tweets.length > 0) {
+    if (!document.querySelector(`#twitter-wjs`)) injectTwitterScript()
 
-    if (
-      typeof twttr !== `undefined` &&
-      window.twttr.widgets &&
-      typeof window.twttr.widgets.load === `function`
-    ) {
-      window.twttr.widgets.load()
-    }
+    // Prevent the Twitter script automatically rendering tweets, and hide them
+    tweets.forEach(tweet => {
+      tweet.className = `twitter-tweet-unrendered`
+      tweet.style.display = `none`
+    })
+
+    window.twttr.ready(() =>
+      tweets.forEach(tweet => {
+        const url = tweet.querySelector(`a[href*="/status/"]`).href
+        const id = url.match(/\/status\/([0-9]+)/)[1]
+
+        window.twttr.widgets
+          .createTweet(id, tweet.parentNode)
+          .then(renderedTweet => {
+            if (renderTweets) {
+              tweet.parentNode.insertBefore(renderedTweet, tweet)
+              renderedTweets.push(renderedTweet)
+            } else {
+              renderedTweet.remove()
+            }
+          })
+      })
+    )
   }
 }
+
+exports.onPreRouteUpdate = function() {
+  // Remove rendered tweets before navigating away, to prevent
+  // memory leaks with React (since it's unaware of them)
+  renderedTweets.forEach(renderedTweet => renderedTweet.remove())
```

Comment:
should clear `renderedTweets` array here?

Reply:
```js
renderedTweets.splice(0).forEach(renderedTweet => renderedTweet.remove())
```